Autorelease to close files and not leak memory after invoking swiftc in SwiftSyntax tool.

davidungar opened the pull request:
<!-- What's in this pull request? -->
Close the output file descriptors after the swiftc invocation. I have testing this code in a different program, but have not tested this in situ. (I don't know how to test this particular tool.)

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
Close files after invoking swiftc in SwiftSyntax tool

davidungar commented:
@swift-ci Please smoke test.

nkcsgexi reviewed (approved):
LGTM! Thank you fixing this.

harlanhaskins commented:
Surprisingly enough, NSPipe doesn't require you to explicitly close the file handle for this, though it seems like it should be required.

https://developer.apple.com/documentation/foundation/nspipe/1414352-filehandleforreading

jrose-apple commented:
Yeah, there should be no reason to do this. That's the advantage of putting file handles in types with deinitializers.

davidungar commented:
Here is a short program that demonstrates the issue, unless I am missing something:

import Foundation

func runSync()  {
    let process = Process()
    process.launchPath = "/bin/echo"
    process.standardOutput = Pipe()
    process.standardError = Pipe()
    
    process.launch()
    process.waitUntilExit()
}

for thousands in 0..<20 {
    for i in 0..<1000 {
        runSync()
    }
    print("finished \(thousands + 1) thousand")
}

jrose-apple commented:
I talked with David about the real culprit: [autorelease pools](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmAutoreleasePools.html)!

```swift
import Foundation

func runSync() {
  let process = Process()
  process.launchPath = "/bin/echo"
  process.standardOutput = Pipe()
  process.standardError = Pipe()

  process.launch()
  process.waitUntilExit()
}

for hundreds in 0..<20 {
  for i in 0..<100 {
    autoreleasepool { // NEW
      runSync()
    }
  }
  print("finished \(hundreds + 1) hundreds")
}
```

That might still be relevant for a long-lived SwiftSyntax tool, but at least this part isn't something to worry about.

Commit:
Use an autoreleasepool instead of explicitly closing the files.

davidungar commented:
@swift-ci Please smoke test

davidungar commented:
Tx, Jordan & Harlan! I have changed the title of this PR and changed the "fix" to use an autoreleasepool and to implicitly close the files. Would you like me to merge this?
